Add ERC: Temporary Approval Extension for ERC-20 (#358)
ethereum/ERCs: ERCS/erc-7674.md

## Motivation

`EIP-1153` allows to use a cheaper way to transiently store allowances.

## Specification

The key words "MUST", "SHOULD", "MAY" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

Compliant contracts MUST implement 1 new function in addition to `ERC-20`:
```solidity
function transientApprove(address spender, uint256 value) public returns (bool success)
```
Call to `transientApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount.

Compliant contracts MUST use the transient storage `EIP-1153` to keep the temporary allowance. For each `owner` and `spender`, the slot MUST be uniquely selected to avoid slot collision. Each slot index SHOULD be derived from the base slot index for transient allowances, `owner` and `spender` addresses. Slot MAY be derived as `keccak256(spender . keccak256(owner . p))` where `.` is concatenation and `p` is `keccak256` from the string uniquely defining transient allowances in the namespace of the implementing contract.

> Compliant contracts MUST add a transient allowance check to the `transferFrom` function. The permanent allowance can only be spent after the temporary allowance has been exhausted.

Reviewer: ```suggestion
Compliant contracts MUST add a transient allowance check to the `transferFrom` function. The permanent allowance SHOULD only be spent after the temporary allowance has been exhausted.
```

> Compliant contracts MUST add a transient allowance to the permanent one when returning the allowed amount to spend in the `allowance` function.

Reviewer: ```suggestion
Compliant contracts MUST add a transient allowance to the permanent one when returning the allowed amount to spend in the `allowance` function. In case the sum of the transient and permanent allowance overflow, `type(uint256).max` SHOULD be returned.
```

---
eip: 7674
title: Temporary Approval Extension for ERC-20

> description: An interface for ERC-20 approvals via transient storage

Reviewer: Transient is not single-block though, its single-transaction. I like the idea, but we would need the wording to accuratly describe what transient does.

Maybe
```suggestion
description: An interface for ephemeral ERC-20 approvals
```

> description: An interface for ERC-20 approvals via transient storage

Reviewer 2: Yeah, 100% agree wth your suggestion 👍🏻

author: Xenia Shape (@byshape), Mikhail Melnik (@ZumZoom), Hadrien Croubois (@Amxx)
discussions-to: https://ethereum-magicians.org/t/add-erc-7674-transient-approval-extension-for-erc-20/19521
status: Draft
type: Standards Track
category: ERC
created: 2024-04-02
requires: 20, 1153
---

## Abstract

> This specification defines the minimum interface required to temporarily approve `ERC-20` tokens for spending within a single transaction.

Reviewer: same comment: "single block" is incorrect.

> This specification defines the minimum interface required to temporarily approve `ERC-20` tokens for spending within a single transaction.

Reviewer 2: Yes. Updated accordingly

> This specification defines the minimum interface required to temporarily approve `ERC-20` tokens for spending within a single transaction.

Reviewer 3: I actually prefer this wording instead of adding extra text in parens:

```suggestion
This specification defines the minimum interface required to temporarily approve `ERC-20` tokens for spending within the same transaction.
```

> This specification defines the minimum interface required to temporarily approve `ERC-20` tokens for spending within a single transaction.

Reviewer 4: ```suggestion
This specification defines the minimum interface required to temporarily approve [ERC-20](./eip-20.md) tokens for spending within a single transaction.
```

> This specification defines the minimum interface required to temporarily approve [ERC-20](./eip-20.md) tokens for spending within the same transaction.

Reviewer 3: ```suggestion
This specification defines the minimum interface required to temporarily approve [ERC-20](./erc-20.md) tokens for spending within the same transaction.
```

## Motivation

> `EIP-1153` allows to use a cheaper way to temporarily store allowances.

Reviewer 4: ```suggestion
[EIP-1153](./eip-1153.md) introduces a cheaper way to temporarily store allowances.
```

## Specification

The key words "MUST", "SHOULD", "MAY" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

> Compliant contracts MUST implement 1 new function in addition to `ERC-20`:

Reviewer 4: Please remove all EIP names in backticks and replace them with links. Backticks are only for inline code.

> Compliant contracts MUST implement 1 new function in addition to [ERC-20](./eip-20.md):

Reviewer 3: ```suggestion
Compliant contracts MUST implement 1 new function in addition to [ERC-20](./erc-20.md):
```

```solidity
function temporaryApprove(address spender, uint256 value) public returns (bool success)
```

> Call to `temporaryApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount.

Reviewer 2: I think this should be updated regardless.

> Call to `temporaryApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount.

Author: If we're talking about the same transaction, I think it's better to mention only `value` since `allowance(owner, spender)` can be spent (or not) later too.

> Call to `temporaryApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount.

Reviewer: ```suggestion
Call to `temporaryApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount. This temporary allowance is to be considered in addition to the normal (persistent) ERC20 allowance, the total value than spender is able to spend during the transaction is thus capped by the sum of the temporary and the normal (pensistent) allowances.
```

> Call to `temporaryApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount.  This temporary allowance is to be considered in addition to the normal (persistent) [ERC-20](./eip-20.md) allowance, the total value that spender is able to spend during the transaction is thus capped by the sum of the temporary and the normal (persistent) allowances.

Reviewer 5: >This temporary allowance is to be considered in addition to the normal (persistent) ERC-20 allowance

This forces an SLOAD of the current allowance, resulting in higher costs than one would want from this feature. I think it should be a temporary allowance override instead.

Reviewer: I disagree. 

If the temporary allowance takes priority (this should be mentionned somewhere if it isn't already), and if the temporary allowance is sufficient to cover the need of some function that consumes allowance (`transferFrom`), then the allowance can be consumed from the temporary part without having to go consume anything in the persistent part.

You only need to lean (and update) the persistent part if the temporary part is not sufficient.

An analogy would be:
- you want to cook rice,
- you have 2kg of rice in your kitchen (fast access - eq to temporary)
- you have 50kg of rice stored in you basement (slow access - eq to persistent)

it is true that:
- you have 52kg available
- to measure how much you have available, you have to check both the kitchen and the basement
- if you need to cook diner, and that diner requires 500g of rice, the kitchen is sufficient, and you don't need to go to the basement. Rice in the basement is available if you need, but you don't need to access it unless there isn't enough in the kitchen for what you are planning to cook.

Reviewer: What do you think of

```suggestion
Call to `temporaryApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount.  This temporary allowance is to be considered in addition to the normal (persistent) [ERC-20](./eip-20.md) allowance, the total value that spender is able to spend during the transaction is thus capped by the sum of the temporary and the normal (persistent) allowances. 

While it SHOULD be possible for a `transferFrom` operation to consume both time of allowance, the consumption of the temporary allowance SHOULD take priority over the consumption of the persistent allowance. Therefore, if the temporary allowance is sufficient for executing a `transferFrom` operation, the persistent allowance SHOULD not be loaded/updated from the storage. Consumption of persistent allowance, which implies storage accesses, SHOULD be performed only if the temporary allowance is not sufficient for the operation being executed.
```

Reviewer 5: Yes, my confusion was because of the effect on the `allowance()` function. If the allowances have to be added, this forces `allowance` to read from both. However, in the context of `transferFrom` it should not be necessary to do this, as you've described.

I am not sure if the effect on `allowance` is something that we should care about.

> Call to `temporaryApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount.  This temporary allowance is to be considered in addition to the normal (persistent) [ERC-20](./eip-20.md) allowance, the total value that spender is able to spend during the transaction is thus capped by the sum of the temporary and the normal (persistent) allowances.

Reviewer 3: ```suggestion
Call to `temporaryApprove(spender, value)` allows `spender` to withdraw within the same transaction from `msg.sender` multiple times, up to the `value` amount.  This temporary allowance is to be considered in addition to the normal (persistent) [ERC-20](./erc-20.md) allowance, the total value that spender is able to spend during the transaction is thus capped by the sum of the temporary and the normal (persistent) allowances.
```

> Compliant contracts MUST use the transient storage `EIP-1153` to keep the temporary allowance. For each `owner` and `spender`, the slot MUST be uniquely selected to avoid slot collision. Each slot index SHOULD be derived from the base slot index for temporary allowances, `owner` and `spender` addresses. Slot MAY be derived as `keccak256(spender . keccak256(owner . p))` where `.` is concatenation and `p` is `keccak256` from the string uniquely defining temporary allowances in the namespace of the implementing contract.

Reviewer 2: > Compliant contracts MUST use the transient storage `EIP-1153` to keep the temporary allowance.

Is this enforceable, though? I agree with @Amxx that "transient" shouldn't be exposed in the ABI or documentation, which makes me think that requiring the use of EIP-1153 is not even needed for the interface and the semantics of `temporaryApprove`.

Author: Suggestion has been applied, MUST was replaced with MAY.

> The storage for the temporary allowances MUST be different to that of the regular allowance. Compliant contracts MAY use the transient storage [EIP-1153](./eip-1153.md) to keep the temporary allowance. For each `owner` and `spender`, the slot MUST be uniquely selected to avoid slot collision. Each slot index SHOULD be derived from the base slot index for temporary allowances, `owner` and `spender` addresses. Slot MAY be derived as `keccak256(spender . keccak256(owner . p))` where `.` is concatenation and `p` is `keccak256` from the string uniquely defining temporary allowances in the namespace of the implementing contract.

Reviewer 4: This is all still implementation detail. You should only specify the externally visible behaviour in the Specification section. You can put recommendations in the Reference Implementation section. How about:

```suggestion
Each temporary allowance MUST persist until the end of the transaction that created it (unless overwritten by another call to `temporaryApprove`.) Each temporary allowance MUST be cleared at the end of the transaction that created it. See [Using Transient Storage](#using-transient-storage) for an example.
```

Then moving this content to the Reference Implementation section under a "Using Transient Storage" heading.

Reviewer: > (unless overwritten by another call to `temporaryApprove` or consumed by a call to `transferFrom`).

> The storage for the temporary allowances MUST be different to that of the regular allowance. Compliant contracts MAY use the transient storage [EIP-1153](./eip-1153.md) to keep the temporary allowance. For each `owner` and `spender`, the slot MUST be uniquely selected to avoid slot collision. Each slot index SHOULD be derived from the base slot index for temporary allowances, `owner` and `spender` addresses. Slot MAY be derived as `keccak256(spender . keccak256(owner . p))` where `.` is concatenation and `p` is `keccak256` from the string uniquely defining temporary allowances in the namespace of the implementing contract.

Reviewer 3: ```suggestion
The storage for the temporary allowances MUST be different to that of the regular allowance. Compliant contracts MAY use the transient storage [EIP-1153](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1153.md) to keep the temporary allowance. For each `owner` and `spender`, the slot MUST be uniquely selected to avoid slot collision. Each slot index SHOULD be derived from the base slot index for temporary allowances, `owner` and `spender` addresses. Slot MAY be derived as `keccak256(spender . keccak256(owner . p))` where `.` is concatenation and `p` is `keccak256` from the string uniquely defining temporary allowances in the namespace of the implementing contract.
```

Compliant contracts MUST add a temporary allowance check to the `transferFrom` function. The permanent allowance SHOULD only be spent after the temporary allowance has been exhausted.

Compliant contracts MUST add a temporary allowance to the permanent one when returning the allowed amount to spend in the `allowance` function. In case the sum of the temporary and permanent allowance overflow, `type(uint256).max` MUST be returned.

> No event is required when setting a temporary allowance, but compliant contracts MAY implement it.

Reviewer 4: ```suggestion
No event is required when setting a temporary allowance, but compliant contracts MAY emit it.
```

> No event is required when setting a temporary allowance, but compliant contracts MAY emit a specific event.

Reviewer 4: You should specify the event, or clarify what "specific" means here. The easiest would be to define the event, but leave it optional.

## Rationale

> The main goal of this standard is to make it cheaper to approve `ERC-20` tokens for a single transaction with minimal interface extension to allow easier integration of a compliant contract into existing infrastructure. This affects the backward compatibility of the `allowance` function. However, the required changes to the `transferFrom` function implementation satisfy the requirement to explicitly authorize the spender to transfer tokens.

Reviewer 4: Comments about the proposal as a whole belong in the Motivation section. The Rationale section should be used to explain choices made within your proposal.

## Backwards Compatibility

All functionality of the `ERC-20` standard is backward compatible except for the `allowance` function.

## Reference Implementation

> TBD

Reviewer 4: If you use an HTML-style comment, the linter will make sure you deal with it before going to review.

```suggestion
TBD <!-- TODO -->
```

## Security Considerations

> The method of deriving slot identifiers to store temporary allowances must avoid collision with other transient storage slots.

Reviewer 2: I acknowledge I'm biased, but why not to suggest using https://eips.ethereum.org/EIPS/eip-7201 right away to derive the slot? It has the characteristics we're looking for imo.

Author: Some may want to go in a different direction. So I think the wording should be flexible here.

> All functionality of the [ERC-20](./eip-20.md) standard is backward compatible except for the `allowance` function.

Reviewer 3: ```suggestion
All functionality of the [ERC-20](./erc-20.md) standard is backward compatible except for the `allowance` function.
```
